climpred/classes.py

```diff
@@ -819,164 +748,104 @@ def verify(
+        if isinstance(reference, str):
+            reference = [reference]
+        elif reference is None:
+            reference = []
+
+        def _verify(
+            hind,
+            verif,
+            hist,
+            reference,
+            metric,
+            comparison,
+            alignment,
+            dim,
+            **metric_kwargs,
+        ):
+            """Interior verify func to be passed to apply func."""
+            result = {}
+            metric, comparison, dim = _get_metric_comparison_dim(
+                metric, comparison, dim, kind=self.kind
+            )
+            forecast, verif = comparison.function(hind, verif, metric=metric)
+            forecast = forecast.rename({'init': 'time'})
+            inits, verif_dates = return_inits_and_verif_dates(
+                forecast, verif, alignment, reference=reference, hist=hist,
+            )
+            metric_over_leads = [
+                _apply_metric_at_given_lead(
+                    verif,
+                    verif_dates,
+                    lead,
+                    hind=forecast,
+                    hist=hist,
+                    inits=inits,
+                    # Ensure apply metric function returns skill and not reference
+                    # results.
+                    reference=None,
+                    metric=metric,
+                    comparison=comparison,
+                    dim=dim,
+                    **metric_kwargs,
+                )
+                for lead in forecast['lead'].data
+            ]
+            skill = xr.concat(metric_over_leads, dim='lead', **CONCAT_KWARGS)
+            skill['lead'] = forecast['lead']
+            result['skill'] = skill
+
+            if reference is not None:
+                if 'historical' in reference:
+                    hist, _ = comparison.function(hist, verif, metric=metric)
+                for r in reference:
+                    metric_over_leads = [
+                        _apply_metric_at_given_lead(
+                            verif,
+                            verif_dates,
+                            lead,
+                            hind=forecast,
+                            hist=hist,
+                            inits=inits,
+                            reference=r,
+                            metric=metric,
+                            comparison=comparison,
+                            dim=dim,
+                            **metric_kwargs,
+                        )
+                        for lead in forecast['lead'].data
+                    ]
+                    ref = xr.concat(metric_over_leads, dim='lead', **CONCAT_KWARGS)
+                    ref['lead'] = forecast['lead']
+                    result[r] = ref
+            return result
 
-        Returns:
-            Dataset of comparison results (if comparing to one observational product),
-            or dictionary of Datasets with keys corresponding to
```

Comment:
Why do you want to have this as a dictionary? Please justify. concat would also work but then it’s also possible that persistence as no member or init dim but initialised skill has

Reply:
I don't see another way to return it. If you have multiple variables, you can't concatenate them together. It's easier to just have a dictionary with the reference written as a key. This also goes for the fact that if you have multiple observations it's a nested dictionary.

So first level is results for a given observation (e.g. ['ERSST', 'HADCRUT']), then the next level is ['skill', 'persistence', 'uninitialized']. Thoughts?

Reply:
I'll merge after your feedback on this

Reply:
Multiple vars can be concated. As datasets. But dict is ok.

Reply:
I would call it initialised skill.

Reply:
I am just a bit afraid that a nested dict is harder to work with. Concatenation datasets are xarray builtin structures.

Reply:
dicts are not as easy to work with compared arrays. That’s my main point

Reply:
With a dict you cannot say dict +273 to convert to kelvin. With an array this works.

Reply:
Yes but it's a dict of xarray datasets. You just extract what you need and do your operations.

E.g.,

```python
results = hindcast.verify(metric='nmae', reference=['persistence', 'historical'])
skill, persistence, historical = results['skill'], results['persistence'], results['historical']

skill += 273
```

Reply:
And for multiple obs, you'd just do `results['ERSST']`, `results['HADCRUT']` to access the dictionary of results from each. Yes you can concatenate multiple variables but you have SST results for persistence, uninitialized, and skill. Do we then rename these to `persistence_SST`, `uninitialized_SST`, etc. to concatenate them? I don't like the idea of renaming variables

Reply:
You see how tedious it gets to work with these dicts. 

With concatenation, you don’t need to rename. Concatenation creates a new dimension and there you have persistence uninitialised and initialised. I think this is more xarray than these dict structures.

Reply:
I think the dicts make it more difficult than it needs to. Show me what’s not possible with concatenation and what’s better with dicts